### PR TITLE
[DOM-52534] AWS - CUR permissions

### PR DESCRIPTION
**Pull request by ddl-juan-cistaro**
### Link to JIRA

[DOM-52534](https://dominodatalab.atlassian.net/browse/DOM-52534)

### What issue does this pull request solve?
Being able to deploy and execute the whole functionality related to CUR and accurate cost report

### What is the solution?
Grant deployer the permissions to create and destroy the infra needed for CUR functionality and also add the permissions to the different resources to perform their tasks

### Testing
I created an environment and then destroyed it to validate that the deployer has the proper permissions and then created another env to validate the functionality.

### CODE REVIEW

File: modules/iam-bootstrap/bootstrap-2.json [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        "glue:DeleteSecurityConfiguration",
+        "glue:GetTags",
+        "glue:CreateCrawler",
+        "glue:CreateTable",
+        "glue:GetTable",
+        "glue:DeleteTable",
+        "glue:GetCrawler",
+        "glue:DeleteCrawler"
+      ],
+      "Resource": "*"
```
Comment (miguelhar) [inline modules/iam-bootstrap/bootstrap-2.json:22]: can we gate resources to specific deploy id? Also keep in mind `iam_bootstrap`   is for deployment infra only.

File: modules/iam-bootstrap/bootstrap-2.json [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    },
+    {
+      "Sid": "AthenaUngated",
+      "Effect": "Allow",
+      "Action": [
+        "athena:*"
+      ],
+      "Resource": "*"
```
Comment (miguelhar) [inline modules/iam-bootstrap/bootstrap-2.json:30]: gate?

File: modules/iam-bootstrap/bootstrap-2.json [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      "Sid": "SQSUngated",
+      "Effect": "Allow",
+      "Action": [
+        "sqs:createqueue",
+        "sqs:deletequeue",
+        "sqs:tagqueue",
+        "sqs:getqueueattributes",
+        "sqs:listqueuetags"
+      ],
+      "Resource": "*"
```
Comment (miguelhar) [inline modules/iam-bootstrap/bootstrap-2.json:42]: gate?

**Review by miguelhar — CHANGES_REQUESTED** [review]

**Comment by miguelhar** [comment]
Needs description on PR

File: modules/infra/submodules/cost-usage-report/lambda.tf [hunk 4]
```diff
@@ -53,6 +59,22 @@ resource "aws_security_group" "lambda" {
   description = "EFS security group"
   vpc_id      = var.network_info.vpc_id
 
+  ingress {
+    description = "Allow inbound TCP traffic exclusively from the crawler initializer lambda."
+    from_port   = 0
```
[the remark this answers is not in the document]
Comment (ddl-juan-cistaro) [reply modules/infra/submodules/cost-usage-report/lambda.tf:64]: We are restricting this to allow inbound and outbound traffic only from the security group is that not enough?

File: modules/iam-bootstrap/main.tf [hunk 5]
```diff
@@ -1,5 +1,5 @@
 locals {
-  iam_policy_paths = length(var.iam_policy_paths) == 0 ? ["${path.module}/bootstrap-0.json", "${path.module}/bootstrap-1.json"] : var.iam_policy_paths
+  iam_policy_paths = length(var.iam_policy_paths) == 0 ? ["${path.module}/bootstrap-0.json", "${path.module}/bootstrap-1.json", "${path.module}/bootstrap-2.json"] : var.iam_policy_paths
```
[the remark this answers is not in the document]
Comment (ddl-juan-cistaro) [reply modules/iam-bootstrap/main.tf:2]: When I was doing some testing the other day I hit the limit, but after that I refactored a little, so I will check if we can stay with only two files

File: modules/infra/submodules/cost-usage-report/lambda.tf (hunk 4 above)
[the remark this answers is not in the document]
Comment (miguelhar) [reply modules/infra/submodules/cost-usage-report/lambda.tf:64]: This is saying that the lambda can take any traffic and go anywhere.

[the remark this answers is not in the document]
Comment (ddl-juan-cistaro) [reply modules/infra/submodules/cost-usage-report/lambda.tf:64]: the `self        = true` assigns itself to the boundary. Here is an example of the result of this config:
Screenshot 2023-12-12 at 12 49 44

File: modules/iam-bootstrap/bootstrap-1.json [hunk 6]
```diff
@@ -199,6 +219,65 @@
...
+        "sqs:getqueueattributes",
+        "sqs:listqueuetags"
+      ],
+      "Resource": [
+        "arn:${partition}:sqs:*:*:${deploy_id}*"
+      ]
     }
   ]
-}
+}
```
Comment (ddl-olsonJD) [inline modules/iam-bootstrap/bootstrap-1.json:283]: add new line

File: modules/iam-bootstrap/bootstrap-2.json [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        "cur:PutReportDefinition",
+        "cur:DescribeReportDefinitions",
+        "cur:DeleteReportDefinition"
+      ],
+      "Resource": [
+        "*"
+      ]
+    }
+  ]
+}
```
Comment (ddl-olsonJD) [inline modules/iam-bootstrap/bootstrap-2.json:57]: nl

**Review by ddl-olsonJD — COMMENTED** [review]
nitpick

**Comment by miguelhar** [comment]
Looks good, we need to fix this warning

```
Warning: Redundant empty provider block
│
│   on ../../../../modules/infra/submodules/cost-usage-report/main.tf line 60:
│   60: provider "aws" { alias = "us-east-1" }
│
│ Earlier versions of Terraform used empty provider blocks ("proxy provider configurations") for child modules to declare their need to be passed a provider configuration by their callers. That approach was ambiguous and is now deprecated.
│
│ If you control this module, you can migrate to the new declaration syntax by removing all of the empty provider "aws" blocks and then adding or updating an entry like the following to the required_providers block of module.infra.module.cost_usage_report:
│     aws = {
│       source = "hashicorp/aws"
│       configuration_aliases = [
│         aws.us-east-1,
│       ]
│     }
```

**Review by ddl-olsonJD — COMMENTED** [review]
Also, let's change the variable name
provision_resources => provision_cost_usage_report?

File: modules/iam-bootstrap/bootstrap-1.json [hunk 8]
```diff
@@ -1,6 +1,26 @@
...
+        "backup:DeleteBackupSelection",
+        "backup:GetBackupSelection",
+        "backup:ListTags",
+        "backup:TagResource",
+        "backup:UntagResource",
+        "backup:DeleteRecoveryPoint"
+      ],
+      "Resource": [
+        "arn:${partition}:backup:*:${account_id}:backup-vault:${deploy_id}-efs",
+        "arn:${partition}:backup:*:${account_id}:backup-plan:*",
```
Comment (miguelhar) [inline modules/iam-bootstrap/bootstrap-1.json:20]: could we restrict to `backup-plan:${deploy_id}*` ?
same for recovery-point

**Review by miguelhar — COMMENTED** [review]
Are we testing `domino_cur.provision_cost_usage_report = true` in `examples/tfvars`

**Review by miguelhar — APPROVED** [review]

**Pull request merged**